src/libraries/System.Private.CoreLib/src/System/Reflection/ModifiedFunctionPointerType.cs

```diff
@@ -0,0 +1,117 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Reflection
+{
+    internal sealed partial class ModifiedFunctionPointerType : ModifiedType
+    {
+        private const string CallingConventionTypePrefix = "System.Runtime.CompilerServices.CallConv";
+
+        private readonly ModifiedType[] _parameterTypes;
+        private readonly ModifiedType _returnType;
+        private Type[]? _callingConventions;
+
+        /// <summary>
+        /// Create a root node.
+        /// </summary>
+        public ModifiedFunctionPointerType(
+            Type functionPointerType,
+            Type[] requiredModifiers,
+            Type[] optionalModifiers,
+            int rootSignatureParameterIndex)
+            : base(functionPointerType, requiredModifiers, optionalModifiers, rootSignatureParameterIndex)
+        {
+            Debug.Assert(functionPointerType.IsFunctionPointer);
+            _returnType = Create(functionPointerType.GetFunctionPointerReturnType(), this, nestedSignatureIndex: 0, nestedSignatureParameterIndex: 0);
```

Comment:
This is more eager than it needs to be. We should be parsing the signature only once somebody asks for it.

Reply:
The parsing is not done here (for modifiers) - this code is just building up the type tree. Pasting for FP is only done when `GetOptional\RequiredModifiers()` is called.

However, currently the code does obtain the root-level modifiers for all types eagerly when GetModifiedXxxType() is called(), but that is done by the field\property\parameter and not here.

Reply:
It would be best to delay parsing of both the type and the modifiers in all situations.

Reply:
(done)